src/main/ngapp/accio/src/app/img-gallery/img-gallery.component.ts

```diff
@@ -132,10 +134,171 @@ export class ImgGalleryComponent implements OnInit {
     dialogRef.afterClosed().subscribe(() => {
       console.log('UpdateImage dialog was closed...');
       console.log('Fetching updated images...');
-      this.getImages();
+      this.loadGalleryImages();
       console.log('Fetched updated images...');
     });
   }
+
+  /** 
+   * Fetch images with parameters specified by class members.
+   */
+  async fetchImages(): Promise<any> {
+    let fetchUrl =
+      '/blobs?' +
+      $.param({
+        'proj-id': this.projectId,
+        'img-name': this.imgName,
+        'mask-name': this.maskName,
+        'with-masks': this.withMasks,
+        'sort-img': this.sortImg,
+        'sort-mask': this.sortMask,
+        tag: this.tag,
+      });
+
+    // fetchUrl returns a list of image objects: 'url', 'name', 'type',
+    // 'utc', 'tags[]', 'masks[]'
+    const response = await fetch(fetchUrl);
+    const imageContent = await response.json();
+    console.log(imageContent);
+    return imageContent;
+  }
+
+  /** 
+   * Fetch images with parameters specified by class members and set
+   * imageArray for display.
+   */
+  async getImages(): Promise<any> {
+    console.log('fetching from projectId: ' + this.projectId);
+
+    this.imageArray = await this.fetchImages();
+
+    if (this.imageArray.length > 0) {
+      this.displayImages = true;
+    }
+  }
+
+  /**
+   * Download a single image or mask.
+   */
+  downloadImage(image: any): void {
+    saveAs(image.url, image.name + '.' + image.type);
+  }
+
+  /** 
+   * Get the Base64 representation of image for zipping.
+   */
+  getBase64String(img: any): string {
+    let canvas = document.createElement('canvas');
+    canvas.width = img.width;
+    canvas.height = img.height;
+    let ctx = canvas.getContext('2d');
+    ctx.drawImage(img, 0, 0);
+    let dataURL = canvas.toDataURL('image');
+    return dataURL.replace(
+      /^data:image\/(png|jpg|gif|pjpeg|bmp|x-icon|svg+xml|webp);base64,/,
+      ''
+    );
+  }
+
+  /**
+   * Download a zip including an image and its masks.
+   */
+  async downloadImageAndItsMasks(image: any): Promise<any> {
+    let zip = new JSZip();
+
+    let img = new Image();
+    img.crossOrigin = 'Anonymous';
+    img.src = image.url;
+    let loadImageString: Promise<string> = new Promise((resolve, reject) => {
+      img.onload = () => resolve(this.getBase64String(img));
+    });
+    zip.file(image.name + '.' + image.type, await loadImageString, {
+      base64: true,
+    });
+
+    this.imgName = image.name;
+    this.maskName = '';
+    this.withMasks = true;
+    this.sortImg = '';
+    this.sortMask = '';
+    this.tag = '';
+
+    let images: any;
+    images = await this.fetchImages();
+
+    for (let i = 0; i < images[0]['masks'].length; i++) {
```

Comment:
Can this be `for (let mask of images[0]['masks'])`?

Reply:
I tried that initially, but typescript kept showing an error that `mask` was of type string. It also prevented me from adding a type assertion for the `let mask` part of that kind of for statement.

Reply:
Update: I tried it again and it didn't give me errors anymore. I guess cloud shell just didn't want me to write shorter for loops the other day.